src/net.cpp

```diff
@@ -799,10 +798,11 @@ size_t CConnman::SocketSendData(CNode& node) const
         assert(data.size() > node.nSendOffset);
         int nBytes = 0;
         {
-            LOCK(node.cs_hSocket);
-            if (node.hSocket == INVALID_SOCKET)
+            LOCK(node.m_sock_mutex);
+            if (!node.m_sock) {
                 break;
-            nBytes = send(node.hSocket, reinterpret_cast<const char*>(data.data()) + node.nSendOffset, data.size() - node.nSendOffset, MSG_NOSIGNAL | MSG_DONTWAIT);
+            }
+            nBytes = node.m_sock->Send(reinterpret_cast<const char*>(data.data()) + node.nSendOffset, data.size() - node.nSendOffset, MSG_NOSIGNAL | MSG_DONTWAIT);
```

Comment:
Probably a silly question: `LOCK(node.m_sock_mutex);` applies even to `node.m_sock->Send(...)` operation but then [here](https://github.com/bitcoin/bitcoin/pull/23604/files#diff-00021eed586a482abdb09d6cdada1d90115abe988a91421851960e26658bed02R1389) one can return contained sockets out of the function.

I understand that `m_sock_mutex` guards only `m_sock` member read & write operations but not necessarily the contained socket. Is that correct or not?

Reply:
That is a good question.

Before this PR the mutex protects the value of the socket variable - that is, to make sure it does not change from e.g. `7` (it is a file descriptor) to `INVALID_SOCKET` in the middle of constructs like:

```cpp
if (socket != INVALID_SOCKET) {
    ... do something with the socket, assuming it is valid ...
}
```

After this PR, it is the same with the exception that the variable is not a bare integer anymore, but a `shared_ptr` and an empty such pointer is used to designate what was designated before by `INVALID_SOCKET`.